features/300-request-verification/request-verification.md

```diff
@@ -2,94 +2,57 @@
 
-`x-skygear-headers-signature`
+## Gateway communication verification
 
-This header contains the signature for headers verification.
+Gateway communication occurs between Gateway and Microservices/Gears.
 
-`x-skygear-body-signature`
+Microservices/Gears are designed to host on Kubernetes. Therefore, for
+communication between Gateway and Microservices/Gears, we use Network Policy to
+ensure that only Gateway may interact with Microservices/Gears.
 
-This header contains the signature for body verification.
+## Web-hook communication verification
 
-## Headers signature generation
+Web-hook communication occurs between Auth Gear and Microservices/external
+services.
 
-1. Derive a byte sequence from all Skygear headers excluding the special headers.
-1. Calculate the uppercase hex digest HMAC-SHA256 of the byte sequence with a shared secret.
-1. Pass this digest as `x-skygear-headers-signature`
+To verify the identity of receiver, we require that the web-hook handler
+endpoint must use HTTPS. Therefore we can assume that identity of the endpoint
+is expected as specified in configuration.
 
-## Headers byte sequence derivation algorithm
+To verify the identity of sender, we use a signature header with shared secret
+to allow receiver to verify the authenticity of request.
 
-1. Let `headers` be the original list of headers.
-1. Let `headers` be `headers` with lowercased header name.
-1. Let `headers` be `headers` whose name starts with `x-skygear-`.
-1. Let `headers` be `headers` excluding the special headers.
-1. Let `headers` be the sorted version `headers` by name alphabetically.
-1. Let `lines` be `headers` with name and value joined with `:`.
-1. Let `content` be `lines` joined with `\r\n`.
-1. If `content` is empty, return.
-1. Return the UTF-8 encoding of `content`.
+### Shared secret
 
-### Example
-
-```python
-import hmac
-import hashlib
-
-
-SPECIAL_HEADERS = [
-    'x-skygear-headers-signature',
-    'x-skygear-body-signature',
-]
-
-# Input
-headers = [
-    ('content-type', 'application/json'),
-    ('content-length', '100'),
-    ('X-Skygear-Auth-userid', 'a'),
-    ('X-SKYGEAR-AUTH-VERIFIED', 'true'),
-    ('x-skygear-auth-disabled', 'false'),
-    ('x-skygear-headers-signature', 'fake'),
-]
-secret = b'secret'
-
-# Headers byte sequence derivation algorithm
-headers = [(n.lower(), v) for n, v in headers]
-headers = [h for h in headers if h[0].startswith('x-skygear-')]
-headers = [h for h in headers if h[0] not in SPECIAL_HEADERS]
-headers = sorted(headers, key=lambda h: h[0])
-lines = [n + ':' + v for n, v in headers]
-content = '\r\n'.join(lines)
-bytes_ = content.encode('utf-8')
-
-
-assert bytes_ == b'x-skygear-auth-disabled:false\r\nx-skygear-auth-userid:a\r\nx-skygear-auth-verified:true'
+The shared secret is set in user configuration per web-hook handler. Developer
+must to set them in configuration to use web-hook feature.
 
+To access the shared secret in microservices, it is recommended to use the
```

Comment:
This brings up a problem. Currently we have no way to expose values (e.g. master key and this shared secret) in tenant config to microservice via environment variable.

Reply:
I intended that the developer would create a secret with the shared secret value, and sync it manually for now.

Reply:
So shall we create a feature issue for this?

Reply:
Opened #379 for tracking this feature.